Manage Package Dialog Refactor

llali opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them.
-->
There's a dialog in notebook extension to manage python packages using pip or anaconda. The dialog only supports installing/uninstalling packages locally. But some notebook users need to install python packages inside BDC or an SQL server instance. The changes in this PR improves the dialog to accept package manage providers from other extensions and give extension ability to use the dialog for different locations. 

- Refactoring Manage Package dialog, adding a model to handle install/uninstall/list packages for different providers.
- Added a label to manage package dialog to show the location of the packages 
- Added two package manage providers notebook currently using (local pip and local conda)
- Added an extension API to register providers to manage packages
- Added unit tests

**screanshot for new "location" label:**
![image](https://user-images.githubusercontent.com/17187348/69664404-294a3500-103d-11ea-8a4f-5d5c8d31c85d.png)

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	/**
+	 * Returns the current provider
+	 */
+	public get currentPackageManageProvider(): IPackageManageProvider {
```
since it can return undefined, the return type should be IPackageManageProvider | undefined

llali replied:
why? any option can be undefined too right? 

---
In reply to: [350545075](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350545075) [](ancestors = 350545075)

Charles-Gagnon replied:
It depends. If strict null checking is enabled then typescript will enforce this kind of behavior - so if you say you return IPackageManageProvider then it could ONLY return something implementing that. So this would be an error since undefined isn't an instanceof IPackageManageProvider.

Now that being said we don't currently have strict null checks enabled for the entire code base. It's been something we (mostly Anthony) have been working on but it's a lot of work fixing up our code to do that.

But we should still try to adhere by those rules whenever we can so it's very clear when something can possibly be undefined since that can save us having to do unnecessary undefined checks in some scenarios.

llali replied:
got it. thanks for clarifying that. 

---
In reply to: [354075140](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r354075140) [](ancestors = 354075140)

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+					if (!this._packageTypes.has(element.packageTarget.location)) {
+						this._packageTypes.set(element.packageTarget.location, []);
+					}
+					this._packageTypes.get(element.packageTarget.location).push(element);
```
(although i'm not sure this case could exist!)

llali replied:
packageManageProviders is a map and doesn't let same providers registered more than once so here for sure there won't be duplicate

---
In reply to: [350543873](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350543873) [](ancestors = 350543873)

llali replied:
I actually let multiple providers with same location and package type, I don't think it breaks anything 

---
In reply to: [350943631](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350943631) [](ancestors = 350943631,350543873)

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	/**
+	 * Installs give packages using current provider
```
nit: given

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	/**
+	 * Returns the location title for current provider
+	 */
+	public getLocationTitle(): string {
```
string | undefined

llali replied:
string can be undefined too no? 

---
In reply to: [350545572](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350545572) [](ancestors = 350545572)

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	/**
+	 * UnInstalls give packages using current provider
```
nit: given

chlafreniere commented on extensions/notebook/src/jupyter/localCondaPackageManageProvider.ts:
```diff
@@ -0,0 +1,72 @@
+import * as constants from '../common/constants';
+
+export class LocalCondaPackageManageProvider implements IPackageManageProvider {
+
```
nit: extra spaces

chlafreniere commented on extensions/notebook/src/jupyter/localCondaPackageManageProvider.ts:
```diff
@@ -0,0 +1,72 @@
+	 */
+	public static ProviderId = 'localhost_Anaconda';
+
+	constructor(private jupyterInstallation: JupyterServerInstallation) {
```
this just reminded me of something that we could noodle on. Should we have a way to get the jupyter server installation from anywhere within the notebook extension (because there should only be one)? I could be wrong but I feel like we need to pass it all over the place.

llali replied:
passing this here gives me the option to add test for this class. which I will in next PR :)

---
In reply to: [350547675](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350547675) [](ancestors = 350547675)

chlafreniere commented on extensions/notebook/src/jupyter/localCondaPackageManageProvider.ts:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	/**
+	 * Installs give packages
```
given

chlafreniere commented on extensions/notebook/src/jupyter/localPipPackageManageProvider.ts:
```diff
@@ -0,0 +1,71 @@
+	/**
+	 * Provider Id for Pip package manage provider
+	 */
+	public static ProviderId = 'localhost_Python';
```
maybe localhost_Pip for more clarity?

chlafreniere commented on extensions/notebook/src/jupyter/localPipPackageManageProvider.ts:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	/**
+	 * Installs give packages
```
given 😄

llali replied:
:((

---
In reply to: [350547959](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350547959) [](ancestors = 350547959)

chlafreniere commented on extensions/notebook/src/types.d.ts:
```diff
@@ -3,13 +3,45 @@
 export interface IExtensionApi {
-	getJupyterController(): JupyterController;
+	getJupyterController(): IJupyterController;
+	registerPackageManager(providerId: string, packageManagerProvider: IPackageManageProvider): void
```
this may be a silly question but i'll ask it anyway... do you have other code in-flight that leverages these new api's? do you anticipate that this set will drastically change in the near future?

llali replied:
I wanted to send the extension that's using in in different PR

---
In reply to: [350548767](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350548767) [](ancestors = 350548767)

llali commented on extensions/notebook/src/common/constants.ts:
```diff
@@ -27,6 +27,8 @@ export const jupyterReinstallDependenciesCommand = 'jupyter.reinstallDependencie
 export const jupyterManagePackages = 'jupyter.cmd.managePackages';
 export const jupyterConfigurePython = 'jupyter.cmd.configurePython';
+export const localhostTitle = 'localhost';
+export const localhostName = 'localhost';
```
I thought we might have different title for it in future but I can change it to only one for now. The dialog support having different name for location name and title, for instance for SQL server the location is mssql but the title can be the current server name for the current cotnext 

---
In reply to: [350535566](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350535566) [](ancestors = 350535566)

Charles-Gagnon replied:
These should be localized too if they're display names. I would suggest splitting localized constants into their own section/file as well to make it clear which ones are which

---
In reply to: [350862859](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350862859) [](ancestors = 350862859,350535566)

llali commented on extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts:
```diff
@@ -32,18 +33,27 @@ export class InstalledPackagesTab {
+
+			// TODO: only supporting single location for now. We should add a drop down for multi locations mode
+			//
+			this.locationComponent = view.modelBuilder.text().withProperties({
```
sure, will add it to the PR 

---
In reply to: [350535849](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350535849) [](ancestors = 350535849)

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts:
```diff
@@ -32,18 +33,27 @@ export class InstalledPackagesTab {
-			}
+
+			// TODO: only supporting single location for now. We should add a drop down for multi locations mode
+			//
```
Better to just create an actual issue in github

llali replied:
I was not sure  if we want to support that at all or not in future, I designed the model to support multiple but didn't want to make decision for UI right now. I'll add a task for it

---
In reply to: [350874329](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350874329) [](ancestors = 350874329)

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+ */
+export class ManagePackageDialogModel {
+
+	public currentProvider: string;
```
The managePackagesDialog has its own currentPkgType - seems like you could remove that and just use that

---
In reply to: [350541682](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350541682) [](ancestors = 350541682)

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	 */
+	public get multiLocationMode(): boolean {
+		return this._options.multiLocations;
+	}
```
_options might be undefined. Would be good to have places in here that use that handle that case appropriately

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	/**
+	 * Returns the package types for given location
+	 */
+	public getPackageTypes(targetLocation?: string): [string, string][] {
```
Agreed, returning an interface that's defined as {providerId: string, packageType: string}[] would be much cleaner

---
In reply to: [350536148](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350536148) [](ancestors = 350536148)

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+			return await provider.listPackages();
+		} else {
+			return Promise.reject('Current Provider is not set');
+		}
```
You can just throw an Error here in async functions

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+			this.currentProvider = providerId;
+		} else {
+			throw Error(`Invalid package type ${providerId}`);
+		}
```
One nit is that it's confusing switching between calling it "package type" and "provider ID". Why not just stick with one consistently throughout?

llali replied:
fixed that. it should only change the provider which has the package type

---
In reply to: [350879426](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350879426) [](ancestors = 350879426)

Charles-Gagnon commented on extensions/notebook/src/jupyter/jupyterController.ts:
```diff
@@ -196,16 +202,31 @@ export class JupyterController implements vscode.Disposable {
+	public registerPackageManagers(providerId: string, packageManageProvider: IPackageManageProvider): void {
+		if (packageManageProvider) {
+			if (!this._packageManageProviders.has(providerId)) {
+				this._packageManageProviders.set(providerId, packageManageProvider);
```
>_packageManageProviders [](start = 9, length = 23)

Seems like it should be an error case (at least log something) if a duplicate provider is registered

Charles-Gagnon commented on extensions/notebook/src/jupyter/jupyterController.ts:
```diff
@@ -196,16 +202,31 @@ export class JupyterController implements vscode.Disposable {
 		}
 	}
 
+	public registerPackageManagers(providerId: string, packageManageProvider: IPackageManageProvider): void {
```
>registerPackageManagers [](start = 8, length = 23)

registerPackageManager

Charles-Gagnon commented on extensions/notebook/src/jupyter/localCondaPackageManageProvider.ts:
```diff
@@ -0,0 +1,72 @@
+	 * @param packages Packages to install
+	 * @param useMinVersion minimum version
+	 */
+	installPackage(packages: IPackageDetails[], useMinVersion: boolean): Promise<void> {
```
>installPackage [](start = 1, length = 14)

installPackages

Charles-Gagnon replied:
and uninstallPackages below

---
In reply to: [350881214](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350881214) [](ancestors = 350881214)

Charles-Gagnon commented on extensions/notebook/src/types.d.ts:
```diff
@@ -3,13 +3,45 @@
+	listPackages(): Promise<IPackageDetails[]>;
+	installPackage(package: IPackageDetails[], useMinVersion: boolean): Promise<void>;
+	uninstallPackage(package: IPackageDetails[]): Promise<void>;
+	canUseProvider(): Promise<boolean>
```
missing semicolon here

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,204 @@
+	/**
+	 * returns the array of target locations
+	 */
+	public targetLocations(): string[] {
```
>targetLocations [](start = 8, length = 15)

Suggest either just making this a property getter or rename to getTargetLocations. Function names should be actions typically

llali commented on extensions/notebook/src/jupyter/localPipPackageManageProvider.ts:
```diff
@@ -0,0 +1,71 @@
+	 * Returns true if the provider can be used
+	 */
+	canUseProvider(): Promise<boolean> {
+		return Promise.resolve(true);
```
yes, I wanted to keep the current behavior in this PR. I can add logic here in another PR

---
In reply to: [350542488](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350542488) [](ancestors = 350542488)

llali commented on extensions/notebook/src/types.d.ts:
```diff
@@ -3,13 +3,52 @@
+	summary: string;
+}
+
+export interface IPackageManageProvider {
```
add comments

chlafreniere commented:
In terms of accessibility, is there anything that need to be concerned about re: the location label?

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,281 @@
+		return undefined;
+	}
+
+	/**
```
nit: spacing

chlafreniere commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,281 @@
+		});
+	}
+
+	/**
```
nit: spacing

llali commented:
I don't think so. still using modelview form which if support accessibility, the new label will support it too 

---
In reply to: [561838509](https://github.com/microsoft/azuredatastudio/pull/8473#issuecomment-561838509) [](ancestors = 561838509)

Charles-Gagnon commented:
@llali You should definitely test that out ahead of time 😄 Let me know if you have any questions how to do that (NVDA is the one I personally use - it's pretty easy to set up)

llali commented:
Charls helped with testing and we fixed one accessibility issue. It was not because of my change I fixed it here anyway

---
In reply to: [561878322](https://github.com/microsoft/azuredatastudio/pull/8473#issuecomment-561878322) [](ancestors = 561878322)

Charles-Gagnon commented on extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts:
```diff
@@ -0,0 +1,281 @@
+		if (provider) {
+			return await provider.getLocationTitle();
+		}
+		return Promise.resolve(undefined);
```
You can just return here. And the return type should be `Promise<string | undefined>` to follow strict null check guidelines

Charles-Gagnon commented on extensions/notebook/src/types.d.ts:
```diff
@@ -3,13 +3,130 @@
+
+	/**
+	 * Returns Package Overview
+	 * @param packageName
```
There's linter failures here for the @param missing descriptions
